### DISCUSSION TITLE
support multi-value model in Prometheus compatible way

### DISCUSSION THREAD

**Discussion by yuanbohan** [GreptimeTeam/greptimedb#3532] [Ideas]
## Purpose

Prometheus is a single-value model, it is to draw a single panel from `query_range` response by default. Since GreptimeDB supports [multi-value model](https://docs.greptime.com/user-guide/concepts/data-model#model), but the compatible `/api/v1/query_range` endpoints can not facilitate multi-value scenarios.

## HTTP response format

- `value` under `values` has multiple value, but the same format as Prometheus when one field is specified in PromQL, or only one column in the table
- `fields` tells the name of each field in the response

### Instant Query response

```json
{
   "metric" : {
      "__name__" : "up",
      "job" : "prometheus",
      "instance" : "localhost:9090"
   },
   "fields": ["field1", "field2", "field3"],
   "values" : [
      [ 1435781430.781, "1", "2", "3" ]
   ]
}
```

### Range Query response

```json
{
   "metric" : {
      "__name__" : "up",
      "job" : "prometheus",
      "instance" : "localhost:9090"
   },
   "fields": ["field1", "field2", "field3"],
   "values" : [
      [ 1435781430.781, "1", "2", "3" ],
      [ 1435781445.781, "1", "2", "3" ],
      [ 1435781460.781, "1", "2", "3" ]
   ]
}
```

## Panel

The recommended way to draw the panels from multi-value response needs to be carefully designed. The potential options maybe:

- overlapped panels (the order of the panels can be changed)
- normal multiple panels (similar as repeat option in Grafana)

**Comment by waynexia** [comment]
cc @ZonaHex @alili for those who have experience in rendering those data.

My one concern is some `field`s might be missing or changed due to computation. Like `max(metrics)` may change the field name to `MAX(metrics)`. So the actual field name will be different from what it was defined and is a runtime variable.
